Port RandAugment from pytorch

GilesBathgate opened the pull request:
I wanted to use RandAugment from the pytorch autoaugment module. https://pytorch.org/vision/main/generated/torchvision.transforms.RandAugment.html

I think this is line for line equivalent to the pytorch code, but maybe needs additional improvements specific to TorchSharp. It works as expected, but depends on my previous PRs

NiklasGustafsson commented:
Three thoughts: 

1) Thank you very much for contributing! I'm impressed by the number of PRs that I returned to after my vacation. :-)

2) When you're adding new functionality, try to see if there's a unit test or two that you can add. When you fix a bug, add a unit test that would have failed with the bug, but passes after the fix.

3) When you add new functionality or fix a bug, add an entry to the RELEASENOTEs.md file, at the top under the next release heading. It makes it easier to highlight the changes when we make a release.

Thanks,

Niklas

GilesBathgate commented:
@NiklasGustafsson Hope you had a nice vacation. I was holding off doing more until I had feedback on the PR, and also haven't found out how to add unit tests at this stage, so I could use some guidance. I implemented these changes for a work demonstration involving a multi-label classifier using vgg16. Since I needed RandAugment, for that demo, I had already implemented the code before contributing.

NiklasGustafsson commented:
As long as you don't add new files, adding unit tests is very straight-forward. Just add a new `public void` function with the `[Fact]` attribute to the end of an existing file (try finding one with related tests) in the 'TestTorchSharp' project, call it something reasonable, and you're off to the races.

Also, for regression tests, the file `TestTensorBugs` is where I have collected most of them, even though they are not all about bugs in the `Tensor` code.

NiklasGustafsson commented:
I wonder if it's as easy as checking (in `Tensor.Dispose()`) whether the associated `grad` Tensor is non-null and disposing it, too...

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,162 @@
+    {
+        private static Tensor apply_op(
+                                Tensor img,
+                                string op_name,
```
TorchSharp convention -- we use enums in situations like this.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,162 @@
+                                InterpolationMode interpolation,
+                                IList<float>? fill)
+        {
+            if (op_name == "ShearX") {
```
Once you have switched to enums, a switch statement may be faster and easier to read.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,162 @@
+        }
+
+
+        internal class RandAugment : ITransform
```
I would place the class in the 'Modules' namespace under 'TorchSharp,' which is the pattern we (have tried to) use elsewhere.

GilesBathgate replied:
I don't think it needs to be exposed under the modules namespace, its `internal`. Other transforms see [Rotate.cs ](https://github.com/dotnet/TorchSharp/blob/main/src/TorchVision/Rotate.cs) follow the pattern I've used here.

NiklasGustafsson replied:
It's easier to find when you're cross-referencing code in VS. 

Not a biggie, but consistency (if you find places where the current code is not) is helpful across a code base.

GilesBathgate replied:
@NiklasGustafsson I think all files under `TorchVision/*.cs` currently! I feel its a separate refactoring issue.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,162 @@
+{
+    public static partial class torchvision
+    {
+        private static Tensor apply_op(
```
Should this just be a private method in the class? Where else is it used?

GilesBathgate replied:
It will be used by `AutoAugment`, `TrivialAugmentWide`, etc...

NiklasGustafsson replied:
Time for a base class with a protected member?

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,162 @@
+            } else if (op_name == "Identity") {
+                // Pass
+            } else {
+                throw new ArgumentException($"The provided operator {op_name} is not recognized.");
```
With an enum, you won't have to throw. Just sayin'... :-)

GilesBathgate replied:
You don't need to convince me. I much prefer enums, or better still class based dispatch. 
the idea was to copy RandAugment, then update it: https://pytorch.org/vision/main/_modules/torchvision/transforms/autoaugment.html#_apply_op

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+                        return F.invert(img);
+                    case opType.Identity:
+                        return img; // Pass
+                    default:
```
I'm surprised the C# compiler doesn't complain about the fact that you have already covered all the possible cases. The body of the default case is, in effect, unreachable code.

NiklasGustafsson replied:
I take that back -- ShearY is not covered. Is that intentional?

GilesBathgate replied:
hmm, No, did it never have, or did I remove it in the refactoring I wonder.

GilesBathgate replied:
@NiklasGustafsson Good catch, I've fixed that now. 

BTW I think c# allows you to cast any int to an enum even when that enum doesn't contain that value, maybe that was an old quirk that would emit a warning now.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+
+            public Tensor call(Tensor img)
+            {
+                var fill = this.fill;
```
I would put a dispose scope declaration in this method, and move the result to the outer scope on return.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+
+            public Tensor call(Tensor img)
+            {
+                var fill = this.fill;
```
Why the local?

GilesBathgate replied:
I've no idea. Pytorch has the same. will remove.

BTW I think its a good approach to copy as close to the original implementation line for line and then refactor, even if the original didn't make sense.

NiklasGustafsson replied:
Definitely. I do the same.

NiklasGustafsson replied:
Oh -- need to say this: for anything with any kind of algorithm in it, I'm always try to be careful to add a link to the original Python file at the top of the file, for attribution as well as reference. 

See TorchVision/Ops/Permute.cs, for example.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+                    var (magnitudes, signed) = op_meta[op_name];
+                    var magnitude = magnitudes.Dimensions > 0 ? magnitudes[this.magnitude] : 0.0f;
+
+                    if (signed && torch.randint(2, new[] { 1 }).ToBoolean())
```
Is it possible to use `stackalloc` instead of `new` here? Probably not.

GilesBathgate replied:
Ah, good call. Never came across `stackalloc` before it must be because I am :older_man:

GilesBathgate replied:
Just the wrong parameters here. Fixed.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+                    if (signed && torch.randint(2, new[] { 1 }).ToBoolean())
+                        magnitude *= -1.0;
+
+                    img = apply_op(img, op_name, magnitude.ToSingle(), interpolation, fill);
```
Since you have to have it as a float, anyway (should it be a double??), isn't it better to do the `ToSingle()` call in the true-branch of the conditional expression on line 122, and leave the local variable a float instead of a Tensor?

GilesBathgate replied:
Yeah, I'll make it double too.

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,172 @@
+                    var (magnitudes, signed) = op_meta[op_name];
+                    var magnitude = magnitudes.Dimensions > 0 ? magnitudes[this.magnitude] : 0.0f;
+
+                    if (signed && torch.randint(2, new[] { 1 }).ToBoolean())
```
Hmmm... I wonder if there isn't a cheaper way for us to get a random scalar than this. I'll have to look into that.

NiklasGustafsson replied:
Created https://github.com/dotnet/TorchSharp/issues/1068 to track

NiklasGustafsson commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,181 @@
+
+        public static partial class transforms
+        {
+            static public ITransform RandAugment(
```
Doc comments, of course.

NiklasGustafsson commented:
@GilesBathgate. All my comments have been addressed, except the need for some unit tests.

GilesBathgate commented:
> @GilesBathgate. All my comments have been addressed, except the need for some unit tests.

Yep. Will make some progress on those over the w/end.

NiklasGustafsson commented:
> > @GilesBathgate. All my comments have been addressed, except the need for some unit tests.
> 
> Yep. Will make some progress on those over the w/end.

Great. While you're at it, add some text at the beginning of the file to point at the Python source code.

Also -- every time you push, a new build starts, but the last one keeps going (it looks like it's been canceled, but it hasn't). It will save some compute resources if you batch up a number of commits into one push.

NiklasGustafsson commented:
And add a line for this PR and the make_grid PR in the RELASENOTES.md file, under the 0.100.4 release notes at the top.

GilesBathgate commented:
@NiklasGustafsson If the you think the unit tests are sufficient, then this PR is completed.

NiklasGustafsson commented:
> @NiklasGustafsson If the you think the unit tests are sufficient, then this PR is completed.

Yup. I'll try to get a clean build, then merge.

GilesBathgate commented on src/TorchVision/AutoAugment.cs:
```diff
@@ -0,0 +1,197 @@
+                    { opType.Identity, (torch.tensor(0.0), false) },
+                    { opType.ShearX, (torch.linspace(0.0, 0.3, num_bins), true) },
+                    { opType.ShearY, (torch.linspace(0.0, 0.3, num_bins), true) },
+                    { opType.TranslateX, (torch.linspace(0.0, 150.0 / 331.0 * image_size.height, num_bins), true) },
```
@NiklasGustafsson I think I spotted a bug here, TranslateX should not be image_size.height, it should be image_size.width, TranslateY vice versa.

GilesBathgate replied:
Fixed as part of #1077
